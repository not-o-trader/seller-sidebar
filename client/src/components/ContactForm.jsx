import React from 'react';
import styled from 'styled-components';
import { formatMoney } from 'accounting';

import FormInput from './FormInput.jsx';

const Wrapper = styled.div`
  background-color: #e6e6e6;
  border: 1px solid #d4d4d4;
  margin-bottom: 15px;
`;

const Inner = styled.div`
  width: 95%;
  margin: 0 auto;
`;

const Label = styled.label`
  display: block;
  margin: 10px 0;
  font-weight: bold;
`;

const TextArea = styled.textarea`
  width: 100%;
  display: block;
  font-size: 14px;
  line-height: 21px;
`;

const Float = styled.div`
  float: left;
`;

const Button = styled.button`
  width: 100%;
  padding: 7px 14px;
  font-size: 14px;
  display: block;
  color: white;
  background-color: rgb(232, 118, 0);
  line-height: 21px;
  box-shadow: rgb(0, 0, 0, 0.1);
  border-bottom-left-radius: 3.5px;
  border-bottom-right-radius: 3.5px;
  border-top-left-radius: 3.5px;
  border-top-right-radius: 3.5px;
`;

const Input = styled.input`
  margin-right: 20px;
`;

class ContactForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      message: '',
      firstName: '',
      lastName: '',
      email: '',
      phone: '',
      enroll: false
    };

    this.handleMessageChange = this.handleChange.bind(this, 'message');
    this.handleFirstNameChange = this.handleChange.bind(this, 'firstName');
    this.handleLastNameChange = this.handleChange.bind(this, 'lastName');
    this.handleEmailChange = this.handleChange.bind(this, 'email');
    this.handlePhoneChange = this.handleChange.bind(this, 'phone');
    this.handleEnrollChange = this.handleChange.bind(this, 'enroll');
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  static getDerivedStateFromProps(props, state) {
    if (state && state.message === '') {
      const { year, make, model, price } = props;
      if (year && make && model && price) {
        const priceStr = formatMoney(price);
        const message = `
        Hi, I'm interested in your ${year} ${make} ${model} listed on Not-O-Trader
        for ${priceStr}. I'd like to get more information about this vehicle and
        confirm its availability.
        `.trim().replace(/[\n ]+/g, ' ');
        return { ...state, message };
      }
    }
    return state;
  }

  handleChange(prop, event) {
    const target = event.target;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    this.setState({ [prop]: value });
  }

  handleSubmit(event) {
    event.preventDefault();
    this.props.send(this.state);
    this.setState({
      message: '',
      firstName: '',
      lastName: '',
      email: '',
      phone: '',
      enroll: false
    });
  }

  render() {
    const placeholder = 'Enter your message here. Maximum 250 characters';
    return (
      <Wrapper>
        <Inner>
          <h2>Email Seller</h2>
          <form onSubmit={this.handleSubmit} autoComplete="on">
            <Label htmlFor="message">Message</Label>
            <TextArea
              name="message"
              maxLength={250}
              rows={8}
              placeholder={placeholder}
              value={this.state.message}
              onChange={this.handleMessageChange}
            >
            </TextArea>
            <FormInput
              name="firstName"
              label="First Name"
              type="text"
              placeholder="First Name"
              value={this.state.firstName}
              autocomplete="given-name"
              handleChange={this.handleFirstNameChange}
            />
            <FormInput
              name="lastName"
              label="Last Name"
              type="text"
              placeholder="Last Name"
              value={this.state.lastName}
              autocomplete="family-name"
              handleChange={this.handleLastNameChange}
            />
            <FormInput
              name="emailAddress"
              label="Email"
              type="email"
              placeholder="Email"
              value={this.state.email}
              autocomplete="email"
              handleChange={this.handleEmailChange}
            />
            <FormInput
              name="phoneNumber"
              label="Phone Number"
              type="tel"
              placeholder=""
              optional="true"
              value={this.state.phone}
              autocomplete="tel"
              handleChange={this.handlePhoneChange}
            />
            <div>
              <Float>
                <Input
                  type="checkbox"
                  name="contactCheckbox"
                  value={this.state.enroll}
                  onChange={this.handleEnrollChange}
                >
                </Input>
              </Float>
              <div>
                Yes, I would like to receive updates, promotions, and discounts
                from Not-O-Trader and their affiliates.
              </div>
            </div>
            <div>
              <p>
                By using this service, you accept the terms of
                our Visitor Agreement
              </p>
            </div>
            <Button
              type="submit"
              id="send-seller-email"
            >
              Submit
            </Button>
          </form>
        </Inner>
      </Wrapper>
    );
  }
}

export default ContactForm;
