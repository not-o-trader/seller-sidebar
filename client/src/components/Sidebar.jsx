import React from 'react';
import axios from 'axios';
import styled from 'styled-components';

import VehiclePrice from './VehiclePrice.jsx';
import SellerInfo from './SellerInfo.jsx';
import ContactForm from './ContactForm.jsx';

const Wrapper = styled.div`
  max-width: 100%;
  font-family: "Roboto", "Helvetica Neue", Helvetica, Arial, sans-serif;
  font-size: 0.875em;
  a:link, a:visited, a:hover, a:active {
    color: blue;
  }
`;

class Sidebar extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
    this.sendSellerEmail = this.sendSellerEmail.bind(this);
  }

  componentDidMount() {
    this.getInfo(1);
  }

  getInfo(vehicleId) {
    axios.get(`/api/seller/${vehicleId}`)
      .then(({ data }) => {
        this.setState(data);
      })
      .catch((error) => {
        console.log('Error getting data:', error);
      });
  }

  sendSellerEmail(data) {
    const extra = {
      vehicleId: this.state.vehicleId,
      sellerEmail: this.state.sellerEmail
    };
    data = Object.assign({}, data, extra);
    axios.post('/api/seller/email', data)
      .then(res => console.log('Successfully sent email'))
      .catch(err => console.error('Error sending email:', err));
  }

  render() {
    return (
      <Wrapper>
        <VehiclePrice price={this.state.vehiclePrice} />
        <hr />
        <SellerInfo
          logo={this.state.sellerLogoUrl}
          name={this.state.sellerName}
          phone={this.state.sellerPhone}
          address={this.state.sellerAddress}
          city={this.state.sellerCity}
          state={this.state.sellerState}
          zip={this.state.sellerZip}
          website={this.state.sellerWebsite}
        />
        <ContactForm
          make={this.state.vehicleMake}
          model={this.state.vehicleModel}
          year={this.state.vehicleModelYear}
          email={this.state.sellerEmail}
          price={this.state.vehiclePrice}
          send={this.sendSellerEmail}
        />
      </Wrapper>
    );
  }
}

export default Sidebar;
