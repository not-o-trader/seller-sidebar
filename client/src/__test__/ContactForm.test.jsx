import React from 'react';
import { shallow, mount } from 'enzyme';

import Sidebar from '../components/Sidebar.jsx';
import ContactForm from '../components/ContactForm.jsx';

describe('<ContactForm />', () => {
  it('renders', () => {
    const wrapper = shallow(
      <ContactForm
        make="Cadillac"
        model="ATS"
        year="2016"
        email="fake@example.com"
        price={40000.00}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('uses its `send` property', () => {
    const send = jest.fn();
    const wrapper = mount(
      <ContactForm
        make="Cadillac"
        model="ATS"
        year="2016"
        email="fake@example.com"
        price={40000.00}
        send={send}
      />
    );

    wrapper.find('button#send-seller-email').simulate('submit');
    expect(send).toHaveBeenCalled();
  });
});
