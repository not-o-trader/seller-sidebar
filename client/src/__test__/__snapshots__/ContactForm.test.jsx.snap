// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactForm /> renders 1`] = `
<styled.div>
  <styled.div>
    <h2>
      Email Seller
    </h2>
    <form
      autoComplete="on"
      onSubmit={[Function]}
    >
      <styled.label
        htmlFor="message"
      >
        Message
      </styled.label>
      <styled.textarea
        maxLength={250}
        name="message"
        onChange={[Function]}
        placeholder="Enter your message here. Maximum 250 characters"
        rows={8}
        value="Hi, I'm interested in your 2016 Cadillac ATS listed on Not-O-Trader for $40,000.00. I'd like to get more information about this vehicle and confirm its availability."
      />
      <FormInput
        autocomplete="given-name"
        handleChange={[Function]}
        label="First Name"
        name="firstName"
        placeholder="First Name"
        type="text"
        value=""
      />
      <FormInput
        autocomplete="family-name"
        handleChange={[Function]}
        label="Last Name"
        name="lastName"
        placeholder="Last Name"
        type="text"
        value=""
      />
      <FormInput
        autocomplete="email"
        handleChange={[Function]}
        label="Email"
        name="emailAddress"
        placeholder="Email"
        type="email"
        value=""
      />
      <FormInput
        autocomplete="tel"
        handleChange={[Function]}
        label="Phone Number"
        name="phoneNumber"
        optional="true"
        placeholder=""
        type="tel"
        value=""
      />
      <div>
        <styled.div>
          <styled.input
            name="contactCheckbox"
            onChange={[Function]}
            type="checkbox"
            value={false}
          />
        </styled.div>
        <div>
          Yes, I would like to receive updates, promotions, and discounts from Not-O-Trader and their affiliates.
        </div>
      </div>
      <div>
        <p>
          By using this service, you accept the terms of our Visitor Agreement
        </p>
      </div>
      <styled.button
        id="send-seller-email"
        type="submit"
      >
        Submit
      </styled.button>
    </form>
  </styled.div>
</styled.div>
`;
